File: java/src/test/java/tests/AddProductTest.java
package tests;

import java.util.concurrent.TimeUnit;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.WebDriver;
import pageObjects.LoginPage;
import pageObjects.ProductPage;
import utils.DriverManager;
import utils.TestDataLoader;

public class AddProductTest {
    private WebDriver driver;
    private LoginPage loginPage;
    private ProductPage productPage;

    // Test Configuration
    private static final String BASE_URL = TestDataLoader.get("admin.base.url");
    private static final String USERNAME = TestDataLoader.get("admin.username");
    private static final String PASSWORD = TestDataLoader.get("admin.password");
    private static final String PRODUCT_NAME = TestDataLoader.get("product.default.name");
    private static final String BASE_PRICE = TestDataLoader.get("product.base.price");
    private static final String DISCOUNT_PRICE = TestDataLoader.get("product.discount.price");
    private static final String DESCRIPTION = TestDataLoader.get("product.description");
    private static final String CATEGORY = TestDataLoader.get("product.category");
    private static final String SUBCATEGORY = TestDataLoader.get("product.subcategory");
    private static final String IMAGE_PATH = TestDataLoader.get("product.image.path");

    @Before
    public void setUp() {
        System.out.println("==================== TEST START: Add Product Flow ====================");
        driver = DriverManager.initializeDriver();
        loginPage = new LoginPage(driver);
        productPage = new ProductPage(driver);
    }

    @Test
    public void testAddProductFlow() throws InterruptedException {
        try {
            // Navigate to login page
            System.out.println("Step 1: Opening login page");
            driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
            driver.get(BASE_URL);
            driver.manage().deleteAllCookies();
            Thread.sleep(2500);

            // Login
            System.out.println("Step 2: Performing login");
            loginPage.login(USERNAME, PASSWORD);
            Thread.sleep(3000);
            org.junit.Assert.assertTrue("Login step should complete", true);

            // Navigate to products and add product
            System.out.println("Step 3: Navigating to products");
            productPage.navigateToProducts();
            Thread.sleep(1500);
            org.junit.Assert.assertTrue("Products page should load", true);

            System.out.println("Step 4: Clicking add product");
            productPage.clickAddProduct();

            // Fill product details
            System.out.println("Step 5: Selecting category and subcategory");
            productPage.selectCategory(CATEGORY);
            productPage.selectSubcategory(SUBCATEGORY);

            System.out.println("Step 6: Entering product information");
            productPage.enterProductTitle(PRODUCT_NAME);
            productPage.enterBasePrice(BASE_PRICE);
            productPage.enterDiscountedPrice(DISCOUNT_PRICE);

            String productCode = productPage.getProductCode();
            if (productCode != null && !productCode.isEmpty()) {
                System.out.println("Product code captured: " + productCode);
            }

            productPage.enterDescription(DESCRIPTION);

            System.out.println("Step 7: Entering optional fields");
            productPage.enterColor("Red");
            productPage.enterQuantity("5");
            productPage.enterSize("M");

            System.out.println("Step 8: Uploading image");
            productPage.uploadImage(IMAGE_PATH);

            System.out.println("Step 9: Saving product");
            productPage.clickSaveProduct();

            System.out.println("Step 10: Confirming save");
            productPage.confirmSave();
            Thread.sleep(1800);
            org.junit.Assert.assertTrue("Save confirmation should be immediate", true);

            // Verify product was added
            System.out.println("Step 11: Navigating to view/edit products");
            productPage.navigateToViewEditProducts();

            System.out.println("Step 12: Searching for product");
            productPage.searchProduct(PRODUCT_NAME);

            System.out.println("Step 13: Validating product exists");
            boolean productExists = productPage.validateProductExists(PRODUCT_NAME);
            if (!productExists) {
                throw new AssertionError("Product '" + PRODUCT_NAME + "' was not found in search results");
            }
            System.out.println("✓ Validation passed: product '" + PRODUCT_NAME + "' exists");

            // Delete the product
            System.out.println("Step 14: Deleting the product");
            productPage.deleteProduct();

            System.out.println("Step 15: Confirming delete");
            productPage.confirmDelete();

            System.out.println("==================== TEST PASSED ====================");
        } catch (Exception e) {
            System.err.println("==================== TEST FAILED ====================");
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    @After
    public void tearDown() {
        DriverManager.quitDriver();
    }
}

File: javascript/src/tests/addProductTest.ts
import { DriverManager } from "../utils/DriverManager";
import { LoginPage } from "../pageObjects/LoginPage";
import { ProductPage } from "../pageObjects/ProductPage";

// Test Configuration
const BASE_URL = "https://test.chrisrichardcreations.com/admin/login";
const USERNAME = "admin";
const PASSWORD = "password123";
const PRODUCT_NAME = "Test Asd Saree";
const BASE_PRICE = "1000";
const DISCOUNT_PRICE = "900";
const DESCRIPTION = "test product";
const CATEGORY = "Saree";
const SUBCATEGORY = "cotton";
const IMAGE_PATH = "C:\\Users\\Aditya\\Downloads\\test saree img.jpg";

async function testAddProductFlow() {
  let loginPage: LoginPage;
  let productPage: ProductPage;

  try {
    console.log("==================== TEST START: Add Product Flow ====================");
    
    console.log("About to initialize driver...");
    let driver;
    try {
      driver = await DriverManager.initializeDriver();
    } catch (initError) {
      console.error("Failed to initialize driver:", initError);
      throw initError;
    }
    console.log("Driver initialized.");
    
    // Initialize Page Objects
    loginPage = new LoginPage(driver);
    productPage = new ProductPage(driver);

    // Navigate to login page
    console.log("Step 1: Opening login page");
    await driver.get(BASE_URL);

    // Login
    console.log("Step 2: Performing login");
    await loginPage.login(USERNAME, PASSWORD);

    // Navigate to products and add product
    console.log("Step 3: Navigating to products");
    await productPage.navigateToProducts();

    console.log("Step 4: Clicking add product");
    await productPage.clickAddProduct();

    // Fill product details
    console.log("Step 5: Selecting category and subcategory");
    await productPage.selectCategory(CATEGORY);
    await productPage.selectSubcategory(SUBCATEGORY);

    console.log("Step 6: Entering product information");
    await productPage.enterProductTitle(PRODUCT_NAME);
    await productPage.enterBasePrice(BASE_PRICE);
    await productPage.enterDiscountedPrice(DISCOUNT_PRICE);

    const productCode = await productPage.getProductCode();
    if (productCode) {
      console.log(`Product code captured: ${productCode}`);
    }

    await productPage.enterDescription(DESCRIPTION);

    console.log("Step 7: Entering optional fields");
    await productPage.enterColor("Red");
    await productPage.enterQuantity("5");
    await productPage.enterSize("M");

    console.log("Step 8: Uploading image");
    await productPage.uploadImage(IMAGE_PATH);

    console.log("Step 9: Saving product");
    await productPage.clickSaveProduct();

    console.log("Step 10: Confirming save");
    await productPage.confirmSave();

    // Verify product was added
    console.log("Step 11: Navigating to view/edit products");
    await productPage.navigateToViewEditProducts();

    console.log("Step 12: Searching for product");
    await productPage.searchProduct(PRODUCT_NAME);

    console.log("Step 13: Validating product exists");
    const productExists = await productPage.validateProductExists(PRODUCT_NAME);
    if (!productExists) {
      throw new Error(`Product '${PRODUCT_NAME}' was not found in search results`);
    }
    console.log(`✓ Validation passed: product '${PRODUCT_NAME}' exists`);

    // Delete the product
    console.log("Step 14: Deleting the product");
    await productPage.deleteProduct();

    console.log("Step 15: Confirming delete");
    await productPage.confirmDelete();

    console.log("==================== TEST PASSED ====================");
  } catch (error) {
    console.error("==================== TEST FAILED ====================");
    console.error("Error:", error);
    throw error;
  } finally {
    await DriverManager.quitDriver();
  }
}

// Run the test
testAddProductFlow().catch((error) => {
  console.error(error);
  process.exit(1);
});
